beginner_source/torchtext_custom_dataset_tutorial.py

```diff
@@ -0,0 +1,383 @@
+with Neural Networks <https://github.com/bentrevett/pytorch-seq2seq/blob/master/1%20-%20Sequence%\
+20to%20Sequence%20Learning%20with%20Neural%20Networks.ipynb>`_) but without using legacy version
+of torchtext.
+
+In this tutorial, we will learn how to:
+
+* Read a dataset
+* Tokenize sentence
+* Apply transforms to sentence
+* Perform bucket batching
+
+Let us assume that we need to prepare a dataset to train a model that can perform English to
+German translation. We will use a tab-delimited German - English sentence pairs provided by
+the `Tatoeba Project <https://tatoeba.org/en>`_ which can be downloaded from this link: `Click
+Here <https://www.manythings.org/anki/deu-eng.zip>`__.
+
+Sentence pairs for other languages can be found in this link:
+
+Link: `https://www.manythings.org/anki/ <https://www.manythings.org/anki/>`__
+"""
+
+# %%
+# Setup
+# -----
+#
+# First, download the dataset, extract the zip, and note the path to the file `deu.txt`.
+#
+# Ensure that following packages are installed:
+#
+# * `Torchdata 0.6.0 <https://pytorch.org/data/beta/index.html>`_ (Installation instructions: `C\
+#   lick here <https://github.com/pytorch/data>`__)
+# * `Torchtext 0.15.0 <https://pytorch.org/text/stable/index.html>`_ (Installation instructions:\
+#   `Click here <https://github.com/pytorch/text>`__)
+# * Spacy (Docs: `Click here <https://spacy.io/usage>`__)
+#
+# Here, we are using `Spacy` to tokenize text. In simple words tokenization means to
+# convert a sentence to list of words. Spacy is a python package used for various Natural
+# Language Processing (NLP) tasks.
+#
+# Download the English and German models from spacy as shown below: ::
+#
+#   python -m spacy download en_core_web_sm
+#   python -m spacy download de_core_news_sm
+
+
+# %%
+# Let us start by importing required modules:
+
+import torchdata.datapipes as dp
+import torchtext.transforms as T
+import spacy
+from torchtext.vocab import build_vocab_from_iterator
+eng = spacy.load("en_core_web_sm") # Load the English model to be used for tokenizing
+de = spacy.load("de_core_news_sm") # Load the German model to be used for tokenizing
+
+# %%
+# Now we will load the dataset
+
+FILE_PATH = 'data/deu.txt'
+dataPipe = dp.iter.IterableWrapper([FILE_PATH])
+dataPipe = dp.iter.FileOpener(dataPipe, mode='rb')
+dataPipe = dataPipe.parse_csv(skip_lines=0, delimiter='\t', as_tuple=True)
+
+# %%
+# In the above code block, we are doing following things:
+#
+# 1. At line 2, we are creating an iterable of filenames
+# 2. At line 3, we pass the iterable to `FileOpener` which then
+#    opens the file in read mode
+# 3. At line 4, we call a function to parse the file, which
+#    again returns an iterable of tuples representing each rows
+#    of the tab-delimited file
+#
+# Data pipes can be thought of something like a dataset object, on which
+# we can perform various operations.
+# Check `this tutorial <https://pytorch.org/data/beta/dp_tutorial.html>`_ for more details on
+# data pipes.
```

Comment:
Change to DataPipes

Reply:
I have made changes as suggested. Also I have added the word "DataPipe" and "DataPipes" in `en-wordlist.txt`.